src/mlpack/core/metrics/non_maximal_supression_impl.hpp

```diff
@@ -0,0 +1,128 @@
+  arma::mat area;
+  if (UseCoordinates)
+  {
+    area = (boundingBoxes.row(2) - boundingBoxes.row(0)) %
+        (boundingBoxes.row(3) - boundingBoxes.row(1));
+  }
+  else
+  {
+    area = (boundingBoxes.row(2)) % (boundingBoxes.row(3));
+  }
+
+  while (sortedIndices.n_elem > 0)
+  {
+    size_t selectedIndex = sortedIndices(sortedIndices.n_elem - 1);
+
+    // Choose the box with the largest probability.
+    selectedIndices.insert_rows(0, arma::uvec(1).fill(selectedIndex));
+
+    // Check if there are other bounding boxes to compare with.
+    if (sortedIndices.n_elem == 1)
+    {
+      break;
+    }
+
+    // Remove the last index.
+    sortedIndices = sortedIndices(arma::span(0, sortedIndices.n_rows - 2),
+        arma::span());
+
+    // Calculate IoU of remaining boxes with the last bounding box with
+    // the highest confidence score.
+    BoundingBoxesType intersectionArea;
+    if (UseCoordinates)
+    {
+      intersectionArea = arma::clamp(arma::clamp(
+          boundingBoxes.submat(arma::uvec(1).fill(2), sortedIndices), DBL_MIN,
+          boundingBoxes(2, selectedIndex)) - arma::clamp(
+          boundingBoxes.submat(arma::uvec(1).fill(0), sortedIndices),
+          boundingBoxes(0, selectedIndex), DBL_MAX), 0.0, DBL_MAX) %
+          arma::clamp(arma::clamp(boundingBoxes.submat(arma::uvec(1).fill(3),
+          sortedIndices), DBL_MIN, boundingBoxes(3, selectedIndex)) -
+          arma::clamp(boundingBoxes.submat(arma::uvec(1).fill(1),
+          sortedIndices), boundingBoxes(1, selectedIndex), DBL_MAX),
+          0.0, DBL_MAX);
+    }
+    else
+    {
+      intersectionArea = arma::clamp(arma::clamp(
+          boundingBoxes.submat(arma::uvec(1).fill(2), sortedIndices) +
+          boundingBoxes.submat(arma::uvec(1).fill(0), sortedIndices), DBL_MIN,
+          boundingBoxes(2, selectedIndex) + boundingBoxes(0, selectedIndex)) -
+          arma::clamp(boundingBoxes.submat(arma::uvec(1).fill(0),
+          sortedIndices), boundingBoxes(0, selectedIndex), DBL_MAX), 0.0,
+          DBL_MAX) % arma::clamp(arma::clamp(
+          boundingBoxes.submat(arma::uvec(1).fill(3), sortedIndices) +
+          boundingBoxes.submat(arma::uvec(1).fill(1),
+          sortedIndices), DBL_MIN, boundingBoxes(3, selectedIndex) +
+          boundingBoxes(1, selectedIndex)) -
+          arma::clamp(boundingBoxes.submat(arma::uvec(1).fill(1),
+          sortedIndices), boundingBoxes(1, selectedIndex), DBL_MAX),
+          0.0, DBL_MAX);
+    }
+
+    BoundingBoxesType calculateIoU = intersectionArea /
+        (area(sortedIndices).t() - intersectionArea + area(selectedIndex));
+
+    sortedIndices = sortedIndices(arma::find(calculateIoU <= threshold));
+  }
+
+  selectedIndices = arma::flipud(selectedIndices);
+}
+
+template<bool UseCoordinates>
+template<typename Archive>
+void NMS<UseCoordinates>::serialize(
+    Archive& ar,
+    const unsigned int /* version */)
+{
+  ar & BOOST_SERIALIZATION_NVP(useCoordinates);
```

Comment:
Since this is a template parameter, this is set once at creation time, so no need to serialize the value.

Reply:
Ahh, makes sense, resolved in the next commit. Thanks a lot.